### ISSUE TITLE
depguard is enabled and fails on v1.53.0

### ISSUE THREAD

**Issue by adamdecaf** [golangci/golangci-lint#3862]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've included all information below (version, config, etc.).
- [X] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

### Description of the problem

Running v1.53.0 of golangci-lint seems to enable `depguard` and fails due to a config error. 

### Version of golangci-lint

<details>

```console
$ golangci-lint --version

v1.53.0
```

</details>

### Configuration file

<details></details>

### Go environment

<details>

```console
$ go version && go env

go version go1.20.4 darwin/amd64
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v

golangci/golangci-lint info checking GitHub for tag 'v1.53.0'
golangci/golangci-lint info found version: 1.53.0 for v1.53.0/darwin/amd64
golangci/golangci-lint info installed ./bin/golangci-lint
STARTING golangci-lint checks
ERRO [linter] depguard: create analyzer: couldn't expand $gostd: could not read GOROOT directory: open src: no such file or directory
must have an Allow and/or Deny package list 
```

</details>

### Code example or link to a public repository

Updating golangci-lint to v1.53.0 in https://github.com/moov-io/infra/blob/master/go/lint-project.sh 

CI run: https://github.com/moov-io/infra/actions/runs/5149020788/jobs/9271486191

**Comment by gandarez** [comment]
There's an open issue to track it but the maintainer hasn't replied yet https://github.com/OpenPeeDeeP/depguard/issues/46

**Comment by adamdecaf** [comment]
Does the upgrade to v2 need to be reverted? https://github.com/golangci/golangci-lint/pull/3795

**Comment by ldez** [comment]
The configuration has changed and it's expected.

We will not revert the upgrade.

**Comment by ldez** [comment]
> It bumped from v1.52.2 to v1.53.0 so no major change was implicitly made

We follow semver but we cannot follow the major version of each linter, then changes inside linter configuration are not considered as breaking.

**Comment by ldez** [comment]
I will try to reproduce the problem locally and find a fix.

**Comment by jhheider** [comment]
It fails something as simple as a `golangci-lint help linters` in an empty directory, where this was a working behavior in the previous version: 

https://github.com/teaxyz/pantry/actions/runs/5148610773/jobs/9271272674

**Comment by adamdecaf** [comment]
I don't have depguard enabled either.

**Comment by ldez** [comment]
I need a reproducible context, then I need information.

```
$ golangci-lint version

$ go env

$ cat .golangci.yml

```

**Comment by adamdecaf** [comment]
`v1.53.0` of golangci-lint fails locally for me inside a Go project. There's no `golangci.yml` file in the project.

```
$ ./bin/golangci-lint version
ERRO [linter] depguard: create analyzer: couldn't expand $gostd: could not read GOROOT directory: open src: no such file or directory
must have an Allow and/or Deny package list 
```

**Comment by ldez** [comment]
I need a reproducible context, then I need information.

```console
$ golangci-lint version

$ go env

$ cat .golangci.yml

```

And a public repo.

**Comment by jhheider** [comment]
```sh
$ mkdir foo
$ cd foo
$ golangci-lint version
ERRO [linter] depguard: create analyzer: couldn't expand $gostd: could not read GOROOT directory: open src: no such file or directory
must have an Allow and/or Deny package list
$ go env
GO111MODULE=""
GOARCH="arm64"
GOBIN=""
GOCACHE="/Users/jacob/Library/Caches/go-build"
GOENV="/Users/jacob/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="arm64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/jacob/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/jacob/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/Users/jacob/.tea/go.dev/v1.20.4"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/Users/jacob/.tea/go.dev/v1.20.4/pkg/tool/darwin_arm64"
GOVCS=""
GOVERSION="go1.20.4"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/dev/null"
GOWORK=""
CGO_CFLAGS="-O2 -g"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-O2 -g"
CGO_FFLAGS="-O2 -g"
CGO_LDFLAGS="-O2 -g"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/t5/180fzzln0k7_0cgpzmkx2_w80000gn/T/go-build2213119517=/tmp/go-build -gno-record-gcc-switches -fno-common"
$ cat .golangci.yml
cat: .golangci.yml: No such file or directory
```

**Comment by ldez** [comment]
I find I know the root of the problem: https://github.com/OpenPeeDeeP/depguard/blob/84475ef6d579f85ae86343f549bd9c267d3b7993/internal/utils/variables.go#L48

The use of `build.Default.GOROOT` cannot work.

**Comment by ldez** [comment]
I'm working on a fix, give me some time.

**Comment by JohnRoesler** [comment]
Thank you @ldez if you need an example on a public repo - I've got it failing here https://github.com/go-co-op/gocron/actions/runs/5149340859/jobs/9272212188

**Comment by adamdecaf** [comment]
I have a public repo which has the failure as well https://github.com/moov-io/infra/pull/271

**Comment by ldez** [comment]
I am still not able to reproduce locally.

I created a simple app just to display `build.Default.GOROOT` and use 2 differents `GOROOT` (one for the build, one for the run) and it works (I'm surprised but it works).

**Comment by ldez** [comment]
What I'm sure about:
```
ERRO [linter] depguard: create analyzer: couldn't expand $gostd: could not read GOROOT directory: open src: no such file or directory
```
the message `open src: no such file or directory` means that the `GOROOT` directory is `src`.

And this directory (`src`) is not your `GOROOT` (GOROOT="/Users/jacob/.tea/go.dev/v1.20.4")

**Comment by adamdecaf** [comment]
I'm on macOS right now and Go is installed via homebrew.

```
go env | grep ROOT
GOROOT="/usr/local/Cellar/go/1.20.4/libexec"
```

**Comment by ldez** [comment]
```console
$ git clone git@github.com:moov-io/infra.git  
Cloning into 'infra'...
remote: Enumerating objects: 9277, done.
remote: Counting objects: 100% (96/96), done.
remote: Compressing objects: 100% (60/60), done.
remote: Total 9277 (delta 42), reused 86 (delta 36), pack-reused 9181
Receiving objects: 100% (9277/9277), 3.93 MiB | 1.45 MiB/s, done.
Resolving deltas: 100% (5751/5751), done.
$ cd infra      
$ golangci-lint run  
$
```

```console
$ git clone git@github.com:go-co-op/gocron.git                  
Cloning into 'gocron'...
remote: Enumerating objects: 1746, done.
remote: Counting objects: 100% (314/314), done.
remote: Compressing objects: 100% (161/161), done.
remote: Total 1746 (delta 218), reused 234 (delta 149), pack-reused 1432
Receiving objects: 100% (1746/1746), 592.91 KiB | 1.54 MiB/s, done.
Resolving deltas: 100% (1158/1158), done.
$ cd gocron
$ golangci-lint run
scheduler.go:1178:11: superfluous-else: if block ends with a break statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary) (revive)
} else {
repeatMap[dayOfMonth]++
}
scheduler_test.go:26:24: unused-parameter: parameter 'i' seems to be unused, consider removing or renaming it as _ (revive)
func (f fakeTime) Unix(i int64, i2 int64) time.Time {
^
scheduler_test.go:2356:54: unused-parameter: parameter 't' seems to be unused, consider removing or renaming it as _ (revive)
func TestScheduler_CheckSetBehaviourBeforeJobCreated(t *testing.T) {
^
```

I'm on Linux amd64.

**Comment by ldez** [comment]
can someone try this command:
```bash
GOROOT=<YOUR REAL GOROOT> golangci-lint run
```

**Comment by adamdecaf** [comment]
I can reproduce this with the following Dockerfile.

```
FROM golang:1.20

RUN wget https://github.com/golangci/golangci-lint/releases/download/v1.53.0/golangci-lint-1.53.0-linux-amd64.tar.gz
RUN tar xf golangci-lint-1.53.0-linux-amd64.tar.gz

RUN go env
RUN /go/golangci-lint-1.53.0-linux-amd64/golangci-lint version
```

Building that fails with the following.

```
 => ERROR [5/5] RUN /go/golangci-lint-1.53.0-linux-amd64/golangci-lint version                                                         0.5s 
------                                                                                                                                      
 > [5/5] RUN /go/golangci-lint-1.53.0-linux-amd64/golangci-lint version:                                                                    
#0 0.468 level=error msg="[linter] depguard: create analyzer: must have an Allow and/or Deny package list"                                  
------
Dockerfile:8
--------------------
   6 |     RUN go env
   7 |     
   8 | >>> RUN /go/golangci-lint-1.53.0-linux-amd64/golangci-lint version
   9 |     
  10 |     # WORKDIR /go/src/
--------------------
ERROR: failed to solve: process "/bin/sh -c /go/golangci-lint-1.53.0-linux-amd64/golangci-lint version" did not complete successfully: exit code: 3
```

**Comment by ldez** [comment]
Thank you, I can reproduce it now.

```console
$ docker run --rm -it golang:1.20-alpine sh                                                                  
/go # wget https://github.com/golangci/golangci-lint/releases/download/v1.53.0/golangci-lint-1.53.0-linux-amd64.tar.gz
Connecting to github.com (140.82.121.4:443)
Connecting to objects.githubusercontent.com (185.199.109.133:443)
saving to 'golangci-lint-1.53.0-linux-amd64.tar.gz'
golangci-lint-1.53.0 100% |********************************************************************************************************************************************************************************************| 9760k  0:00:00 ETA
'golangci-lint-1.53.0-linux-amd64.tar.gz' saved
/go # tar xf golangci-lint-1.53.0-linux-amd64.tar.gz
/go # go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/root/.cache/go-build"
GOENV="/root/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.20.4"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="0"
GOMOD="/dev/null"
GOWORK=""
CGO_CFLAGS="-O2 -g"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-O2 -g"
CGO_FFLAGS="-O2 -g"
CGO_LDFLAGS="-O2 -g"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -fno-caret-diagnostics -Qunused-arguments -Wl,--no-gc-sections -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build3482902365=/tmp/go-build -gno-record-gcc-switches"
/go # /go/golangci-lint-1.53.0-linux-amd64/golangci-lint version
ERRO [linter] depguard: create analyzer: must have an Allow and/or Deny package list 
/go # GOROOT=$(go env GOROOT) /go/golangci-lint-1.53.0-linux-amd64/golangci-lint version
golangci-lint has version 1.53.0 built with go1.20.4 from 9f2528a3 on 2023-06-01T19:43:45Z

```

If I set the `GOROOT` explicitly it works :thinking:

**Comment by ldez** [comment]
More information: the problem appears even if you don't use `depguard` because the function is used inside a global variable.

Another example of why I hate global variables :cry:

**Comment by ldez** [comment]
For me, the problem is not related to build/runtime but to when the global variable is filled.

**Comment by adamdecaf** [comment]
Correct, but most users of golangci-lint probably don't have a GOROOT of `/opt/hostedtoolcache/go/1.20.4/x64/` so they'll see broken builds. I think we need to patch depguard or disable its initialization in golangci-lint until upstream is fixed.

**Comment by ldez** [comment]
your example has a problem because you are changing 2 things:
- the value of `GOROOT`
- how GOROOT is set.

So I don't think `/opt/hostedtoolcache/go/1.20.4/x64/` is used. (the error message says that GOROOT is `src` because inside Go the path `src` should be append or use during the build of field `build.Default.GOROOT`.)

To see what I say about GOROOT:
```console
/go/sandbox # go env | grep GOROOT
GOROOT="/usr/local/go"
/go/sandbox # env | grep GOROOT
/go/sandbox # 
```

**Comment by adamdecaf** [comment]
I thought `$GOROOT/src` refers to the Go source tree that's pulled down for the toolchain to build with. `src` looks to be a relative path off of GOROOT. 

```
root@bb48f4e764af:/go# ls -l /opt/hostedtoolcache/go/1.20.4/x64 | grep src
lrwxrwxrwx 1 root root 17 Jun  1 22:10 src -> /usr/local/go/src
```

```
root@bb48f4e764af:/go# ls -l /opt/hostedtoolcache/go/1.20.4/x64/src/
total 336
-rw-r--r--  1 root root   553 Apr 26 15:31 Make.dist
-rw-r--r--  1 root root  2163 Apr 26 15:31 README.vendor
-rwxr-xr-x  1 root root   407 Apr 26 15:31 all.bash
-rw-r--r--  1 root root   758 Apr 26 15:31 all.bat
-rwxr-xr-x  1 root root   388 Apr 26 15:31 all.rc
drwxr-xr-x  4 root root  4096 Apr 26 15:34 archive
drwxr-xr-x  2 root root  4096 Apr 26 15:34 arena
-rwxr-xr-x  1 root root  2214 Apr 26 15:31 bootstrap.bash
drwxr-xr-x  2 root root  4096 Apr 26 15:34 bufio
-rwxr-xr-x  1 root root  1874 Apr 26 15:31 buildall.bash
drwxr-xr-x  2 root root  4096 Apr 26 15:34 builtin
drwxr-xr-x  2 root root  4096 Apr 26 15:34 bytes
-rwxr-xr-x  1 root root   518 Apr 26 15:31 clean.bash
-rw-r--r--  1 root root   600 Apr 26 15:31 clean.bat
-rwxr-xr-x  1 root root   380 Apr 26 15:31 clean.rc
drwxr-xr-x 25 root root  4096 Apr 26 15:34 cmd
-rw-r--r--  1 root root  1519 Apr 26 15:32 cmp.bash
drwxr-xr-x  8 root root  4096 Apr 26 15:34 compress
drwxr-xr-x  5 root root  4096 Apr 26 15:34 container
drwxr-xr-x  2 root root  4096 Apr 26 15:34 context
drwxr-xr-x 23 root root  4096 Apr 26 15:34 crypto
...
```

**Comment by ldez** [comment]
In conclusion, the problem is the fact to use `build.Default.GOROOT`: this value is filled only if the `GOROOT` is defined through an env var.
A side problem is the use of a global variable to initialize the value.

**Comment by ldez** [comment]
You confirmed my previous comment https://github.com/golangci/golangci-lint/issues/3862#issuecomment-1572872961

**Comment by ldez** [comment]
It's a global variable inside an internal package, I'm sad :cry: 
Then I cannot override it :cry:

**Comment by adamdecaf** [comment]
Can we catch and throw this error only if `depguard` is enabled? I think depguard v2 needs to be reverted or disabled until an upstream fix can be released.

**Comment by ldez** [comment]
> Can we catch and throw this error only if depguard is enabled? 

no because it's a global variable :cry:

**Comment by ldez** [comment]
FYI the `src` comes from here: https://github.com/OpenPeeDeeP/depguard/blob/84475ef6d579f85ae86343f549bd9c267d3b7993/internal/utils/variables.go#L48

**Comment by ldez** [comment]
I made a mistake: the global is used to get the "expander" but the error comes from the call of the "expander" inside the constructor the linter.
Then I think I can catch the error.
But it will not fix the problem for depguard users.

**Comment by ldez** [comment]
Your [PR](https://github.com/OpenPeeDeeP/depguard/pull/48) doesn't fix the problem.

It's the opposite:
- if the env var `GOROOT` is empty then `build.Default.GOROOT` is empty
- if the env var `GOROOT` is **not** empty then `build.Default.GOROOT` is **not** empty

**Comment by ldez** [comment]
A dirty fix is:
```go
	goroot := build.Default.GOROOT
	
	if goroot == "" {
		cmd := exec.Command("go", "env", "GOROOT")
		output, err := cmd.CombinedOutput()
		if err != nil {
			return nil ,fmt.Errorf("get GOROOT: %v", err)
		}
		
		goroot =  string(bytes.TrimSpace(output))
	}
	
	
	root := path.Join(goroot, "src")
```

**Comment by ldez** [comment]
I feel less alone with "dirty" idea when I see that https://github.com/golang/tools/blob/86c93e8732cce300d0270bce23117456ce92bb17/cmd/godoc/goroot.go#L15-L30

**Comment by ldez** [comment]
What I will do:
- create a fork inside the golangci org
- use the "dirty" fix
- open a PR based on this fork on the depguard repo

**Comment by adamdecaf** [comment]
I've updated https://github.com/OpenPeeDeeP/depguard/pull/48 from that golang/tools link.

**Comment by ldez** [comment]
Not super fair, I did the analysis and I find the fix, at least add me as a co-author.
But anyway, I think your PR will not be merged quickly so I do plan to be able to create a release quickly.

**Comment by adamdecaf** [comment]
Fair point. I added you on the commit.

**Comment by ldez** [comment]
I created a PR to fix the problem inside golangci-lint https://github.com/golangci/golangci-lint/pull/3866

I will create a bugfix release when the PR will be reviewed and approved by the other maintainers.

**Comment by deitch** [comment]
I know this is a year old, but picking it up now.

If I understand correctly, after this fix, if you do not have depguard enabled, everything should work. And, indeed, running `golangci-lint run ./... --disable depguard` works.

If you use `depguard` it remains broken? I had thought depguard was now built in with a fix in a fork based on [this comment](https://github.com/golangci/golangci-lint/issues/3862#issuecomment-1572907138), and so it should be fixed for depguard-enabled as well? Or is that fork just to fix the "breaks everything else even if not enabled"?

**Comment by ldez** [comment]
`depguard` has been updated since that issue.

We don't use a fork now and you need to disable it.

You should update your configuration: https://golangci-lint.run/usage/linters/#depguard

**Comment by deitch** [comment]
Thanks @ldez .
